l_dim: 2
N_dim: 2